File: lab5c#/Notification.cs
using System;

namespace SmartHomeSystem
{
    public class Notification
    {
        public string NotificationId { get; private set; }
        public NotificationType Type { get; set; }
        public string Message { get; set; }
        public DateTime Timestamp { get; private set; }
        public bool IsRead { get; private set; }
        public object RelatedObject { get; set; }
        public string Sender { get; set; }

        public Notification(string notificationId, NotificationType type, string message,
                          object relatedObject = null, string sender = "Система")
        {
            NotificationId = notificationId;
            Type = type;
            Message = message;
            Timestamp = DateTime.Now;
            IsRead = false;
            RelatedObject = relatedObject;
            Sender = sender;
        }

        public void MarkAsRead()
        {
            IsRead = true;
        }

        public void DisplayInfo()
        {
            string typeString = "";
            switch (Type)
            {
                case NotificationType.INFO: typeString = "[ИНФО]"; break;
                case NotificationType.WARNING: typeString = "[ПРЕДУПР]"; break;
                case NotificationType.ALERT: typeString = "[ТРЕВОГА]"; break;
            }

            string readStatus = IsRead ? "✓" : "✗";
            Console.WriteLine($"{readStatus} {Timestamp:HH:mm:ss} {typeString} {Message}");
        }

        public string Serialize()
        {
            return $"{NotificationId}|{(int)Type}|{Message}|{Timestamp}|{IsRead}|{Sender}";
        }

        public static Notification Deserialize(string data)
        {
            var parts = data.Split('|');
            if (parts.Length != 6) return null;

            return new Notification(
                parts[0],
                (NotificationType)int.Parse(parts[1]),
                parts[2]
            )
            {
                Timestamp = DateTime.Parse(parts[3]),
                IsRead = bool.Parse(parts[4]),
                Sender = parts[5]
            };
        }

        public override string ToString()
        {
            return $"{Timestamp:HH:mm:ss} - {Message}";
        }
    }
}

File: lab6c++/Notification.cpp
#include "notification.hpp"
#include "device.hpp"
#include "automationScenario.hpp"
#include <iostream>
#include <ctime>
#include <sstream>

Notification::Notification(const std::string& id, NotificationType notificationType,
    const std::string& msg)
    : notificationId(id), type(notificationType), message(msg),
    relatedDevice(nullptr), relatedScenario(nullptr),
    timestamp(std::time(nullptr)), isRead(false) {}

Notification::~Notification() {
}

void Notification::send() {
    std::cout << "Уведомление отправлено: " << message << std::endl;
}

void Notification::markAsRead() {
    isRead = true;
    std::cout << "Уведомление отмечено как прочитанное" << std::endl;
}

int Notification::getNotificationPriority() const {
    switch (type) {
    case NotificationType::ALERT: return 3;
    case NotificationType::WARNING: return 2;
    case NotificationType::INFO: return 1;
    default: return 0;
    }
}

void Notification::displayInfo() const {
    std::cout << "[" << notificationId << "] ";
    switch (type) {
    case NotificationType::ALERT: std::cout << "[ТРЕВОГА] "; break;
    case NotificationType::WARNING: std::cout << "[ПРЕДУПРЕЖДЕНИЕ] "; break;
    case NotificationType::INFO: std::cout << "[ИНФО] "; break;
    }
    std::cout << message;
    if (relatedDevice) {
        std::cout << " (Устройство: " << relatedDevice->getName() << ")";
    }
    if (relatedScenario) {
        std::cout << " (Сценарий: " << relatedScenario->getName() << ")";
    }
    std::cout << " [" << (isRead ? "Прочитано" : "Новое") << "]" << std::endl;
}

std::string Notification::getNotificationId() const {
    return notificationId;
}

NotificationType Notification::getType() const {
    return type;
}

std::string Notification::getMessage() const {
    return message;
}

bool Notification::getIsRead() const {
    return isRead;
}

void Notification::setRelatedDevice(std::shared_ptr<Device> device) {
    relatedDevice = device;
}

void Notification::setRelatedScenario(std::shared_ptr<AutomationScenario> scenario) {
    relatedScenario = scenario;
}

std::string Notification::serialize() const {
    std::stringstream ss;
    ss << notificationId << "|" << static_cast<int>(type) << "|" << message << "|"
        << timestamp << "|" << isRead;
    return ss.str();
}

std::unique_ptr<Notification> Notification::deserialize(const std::string& data) {
    std::stringstream ss(data);
    std::string id, typeStr, message, timestampStr, isReadStr;

    std::getline(ss, id, '|');
    std::getline(ss, typeStr, '|');
    std::getline(ss, message, '|');
    std::getline(ss, timestampStr, '|');
    std::getline(ss, isReadStr, '|');

    auto notification = std::make_unique<Notification>(
        id,
        static_cast<NotificationType>(std::stoi(typeStr)),
        message
    );

    if (isReadStr == "1") notification->markAsRead();
    return notification;
}